File: group_anagrams_test.go
package leetcode

import (
	"fmt"
	"sort"
	"strings"
	"testing"
)

type testCase struct {
	input    []string
	expected [][]string
}

func testCases() []testCase {
	return []testCase{
		testCase{
			input: []string{"eat", "tea", "tan", "ate", "nat", "bat"},

			expected: [][]string{
				[]string{"ate", "eat", "tea"},
				[]string{"bat"},
				[]string{"nat", "tan"},
			},
		},
		testCase{
			input:    []string{""},
			expected: [][]string{[]string{""}},
		},
		testCase{
			input:    []string{"a"},
			expected: [][]string{[]string{"a"}},
		},
	}
}

func TestGroupAnagrams(t *testing.T) {
	for i, tc := range testCases() {
		t.Run(fmt.Sprintf("Test#%d", i+1), func(t *testing.T) {
			act := groupAnagrams(tc.input)
			sortStringSliceSlice(act)

			if !equalStringSliceSlices(act, tc.expected) {
				t.Errorf("groupAnagrams: got %v (sorted), expected %v", act, tc.expected)
			}
		})
	}
}

func sortStringSliceSlice(s [][]string) {
	for _, v := range s {
		sort.Slice(v, func(i, j int) bool { return v[i] < v[j] })
	}

	sort.Slice(s, func(i, j int) bool { return strings.Join(s[i], "") < strings.Join(s[j], "") })
}

func equalStringSliceSlices(a, b [][]string) bool {
	if len(a) != len(b) {
		return false
	}

	for i := range a {
		if !equalStringSlices(a[i], b[i]) {
			return false
		}
	}

	return true
}

func equalStringSlices(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}

	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}

	return true
}

File: group_anagrams_test.rb
# frozen_string_literal: true

require "set"
require "minitest/autorun"
require_relative "group_anagrams"

module ArrayRefinements
  refine Array do
    def deep_clone
      map(&:deep_clone)
    end
  end

  refine String do
    alias_method :deep_clone, :clone
  end
end

class TestGroupAnagrams < Minitest::Test
  TestCase = Struct.new(:strs, :expected, keyword_init: true)

  test_cases = [
    TestCase.new(
      strs: ["eat", "tea", "tan", "ate", "nat", "bat"],
      expected: [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
    ),
    TestCase.new(
      strs: [""],
      expected: [[""]],
    ),
    TestCase.new(
      strs: ["a"],
      expected: [["a"]],
    ),
  ]

  test_cases.each_with_index do |test_case, i|
    define_method("test_case_#{i+1}") do
      actual = group_anagrams(test_case.strs)
      assert_deep_same_elements(test_case.expected, actual)
    end
  end

  using ArrayRefinements

  def assert_deep_same_elements(expected, actual)
    expected = expected.deep_clone
    expected.map!(&:sort!)
    expected = expected.to_set

    actual = actual.deep_clone
    actual.map!(&:sort!)
    actual = actual.to_set

    assert_equal expected, actual
  end
end
